### PR TITLE
Use struct serialization

### PR DESCRIPTION
**Pull request by yorhodes**
Relies on https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/1552

### CODE REVIEW

File: lib/ConfigLib.sol [hunk 1]
```diff
@@ -4,213 +4,108 @@ import "../lib/forge-std/src/Script.sol";
...
-                json,
-                string.concat(chainName, string.concat(prefix, contractName))
-            )
-        returns (bytes memory result) {
-            address parsedAddr = abi.decode(result, (address));
-            return parsedAddr == address(0x20) ? address(0) : parsedAddr;
-        } catch {
-            return address(0);
-        }
+    function readCore(Vm vm, string memory chainName) internal view returns (Core memory) {
```
Comment (asaj) [inline lib/ConfigLib.sol]: So much nicer

File: lib/ConfigLib.sol [hunk 2]
```diff
@@ -4,213 +4,108 @@ import "../lib/forge-std/src/Script.sol";
...
+import {IInterchainSecurityModule} from "@hyperlane-xyz/core/interfaces/IInterchainSecurityModule.sol";
 
 library ConfigLib {
     using stdJson for string;
     using BytesLib for bytes;
 
-    struct HyperlaneDomainConfig {
-        string chainName;
-        uint32 domainId;
+    struct Core {
```
Comment (asaj) [inline lib/ConfigLib.sol:19]: Keys must also be alphabetically sorted, or no?

File: lib/ConfigLib.sol [hunk 3]
```diff
@@ -4,213 +4,108 @@ import "../lib/forge-std/src/Script.sol";
...
+        uint32 domain;
+        IInterchainSecurityModule ism;
```
Comment (asaj) [inline lib/ConfigLib.sol]: nit: defaultIsm

File: scripts/DeployCore.s.sol [hunk 4]
```diff
@@ -7,31 +7,22 @@ import {ConfigLib} from "../lib/ConfigLib.sol";
...
-
-        // Write the output to disk
-        // TODO: Only write if broadcast is on...
-        config.write(vm);
-        // TODO: Shouldn't overwrite startblock...
-        config.writeAgentConfig(vm, startBlock);
+
+        ConfigLib.Core memory coreConfig = ConfigLib.readCore(vm, local);
+        // vm.startBroadcast(deployerPrivateKey);
+        // uint256 startBlock = block.number;
```
Comment (asaj) [inline scripts/DeployCore.s.sol]: We need this for the agent config

File: scripts/DeployCore.s.sol [hunk 5]
```diff
@@ -7,31 +7,22 @@ import {ConfigLib} from "../lib/ConfigLib.sol";
...
-// TODO: Verification
 contract DeployCore is Script {
-    using ConfigLib for ConfigLib.HyperlaneDomainConfig;
-    using CheckLib for ConfigLib.HyperlaneDomainConfig;
-    using DeployLib for ConfigLib.HyperlaneDomainConfig;
+    using DeployLib for ConfigLib.Core;
+    using CheckLib for DeployLib.Core;
 
-    function run() public {
+    function run() public returns (DeployLib.Core memory) {
```
Comment (asaj) [inline scripts/DeployCore.s.sol]: Doesn't write the config?

File: lib/CheckLib.sol [hunk 6]
```diff
@@ -3,103 +3,60 @@ pragma solidity ^0.8.17;
...
-            ConfigLib.MultisigIsmDomainConfig memory domain = config.domains[i];
+            MultisigIsm.DomainConfig memory domainConfig = config.domains[i];
             require(
-                domain.threshold == ism.threshold(domain.domainId),
-                string.concat(
-                    "Default MultisigIsm threshold misconfigured for",
-                    domain.chainName
-                )
+                domainConfig.threshold == ism.threshold(domainConfig.domain),
+                "Threshold misconfigured for"
```
Comment (asaj) [inline lib/CheckLib.sol]: For what?

**Comment by asaj** [comment]
Why do we need to initialize the MultisigIsm in the constructor in order to make these changes?

**Comment by yorhodes** [comment]
> Why do we need to initialize the MultisigIsm in the constructor in order to make these changes?

it doesn't, just figured constructor changes aren't a big deal

File: lib/ConfigLib.sol (hunk 2 above)
Comment (yorhodes) [reply lib/ConfigLib.sol:19]: uhhhh, they are?

**Comment by asaj** [comment]
It's less the constructor changes and more the removal of external functions

**Comment by asaj** [comment]
Also flagging that the permissionless deployment docs will need updates

File: lib/ConfigLib.sol (hunk 2 above)
Comment (asaj) [reply lib/ConfigLib.sol:19]: Yeah I just meant include the same comment as you did elsewhere if it's required here too

File: README.md [hunk 7]
```diff
@@ -52,7 +52,7 @@ export RPC_URL=YOUR_CHAIN_RPC_URL
 # Used to configure the default MultisigIsm.
 export REMOTES=ethereum,polygon,avalanche,celo,arbitrum,optimism,bsc,moonbeam
 
-forge script scripts/DeployCore.s.sol --broadcast --rpc-url $RPC_URL --private-key $PRIVATE_KEY
```
Comment (asaj) [inline README.md:55]: Why remove?

File: foundry.toml [hunk 8]
```diff
@@ -8,7 +8,10 @@ allow_paths = ["../node_modules"]
 solc = '0.8.17'
 optimizer = true
 optimizer_runs = 999_999
-fs_permissions = [{ access = "read-write", path = "./"}]
+fs_permissions = [
+    { access = "read", path = "./config" },
+    { access = "read-write", path = "./deployments" }
```
Comment (asaj) [inline foundry.toml:13]: What are these?

File: lib/CheckLib.sol [hunk 9]
```diff
@@ -3,103 +3,116 @@ pragma solidity ^0.8.17;
...
+import {IInterchainSecurityModule} from "@hyperlane-xyz/core/interfaces/IInterchainSecurityModule.sol";
 import {TransparentUpgradeableProxy} from "@hyperlane-xyz/core/contracts/upgrade/TransparentUpgradeableProxy.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {ProxyAdmin} from "@openzeppelin/contracts/proxy/transparent/ProxyAdmin.sol";
 
 library CheckLib {
-    function check(
-        ConfigLib.HyperlaneDomainConfig memory config,
-        ConfigLib.MultisigIsmConfig memory ismConfig
+    struct AddressAssertion {
```
Comment (asaj) [inline lib/CheckLib.sol:15]: Nice

File: lib/ConfigLib.sol [hunk 10]
```diff
@@ -3,214 +3,95 @@ pragma solidity ^0.8.17;
...
+    //     uint32 domain
+    // ) internal view returns (Agent memory) {
+    //     return
+    //         Agent({
+    //             addresses: core,
+    //             domain: domain,
+    //             finalityBlocks: 1,
+    //             index: Index({from: block.number}),
+    //             rpcStyle: "ethereum",
+    //             url: "" // TODO: vm.rpcUrl(chainName)
```
Comment (asaj) [inline lib/ConfigLib.sol:94]: Fine to leave this blank for now

File: lib/ConfigLib.sol [hunk 11]
```diff
@@ -3,214 +3,95 @@ pragma solidity ^0.8.17;
...
-                contracts,
-                "create2Factory",
-                address(config.create2)
-            )
-            .write(
-                "./config/networks.json",
-                string.concat(".", config.chainName, ".contracts")
-            );
-    }
+    // function buildAgentConfig(
```
Comment (asaj) [inline lib/ConfigLib.sol:83]: What do we need to do to get this back? It's useful (for now) for folks spinning up agents

File: lib/DeployLib.sol [hunk 12]
```diff
@@ -12,133 +11,49 @@ import {Create2Factory} from "@hyperlane-xyz/core/contracts/Create2Factory.sol";
...
-            config.igp = InterchainGasPaymaster(address(proxy));
-            config.igp.transferOwnership(config.owner);
-        } else {
-            console.log(
-                "Found InterchainGasPaymaster at address %s, skipping deployment",
-                address(config.igp)
-            );
-        }
-    }
+    function deploy(ConfigLib.Core memory config) internal returns (Core memory) {
```
Comment (asaj) [inline lib/DeployLib.sol:24]: I see you removed incremental deploys. Nbd as we don't really use them here, just curious to get your thoughts

File: scripts/DeployCore.s.sol [hunk 13]
```diff
@@ -3,31 +3,35 @@ pragma solidity ^0.8.17;
...
+    function run() public returns (DeployLib.Core memory core) {
+        string memory local = vm.envString("LOCAL");
 
-        config.deploy(ismConfig);
-        config.check(ismConfig);
+        ConfigLib.Core memory coreConfig = ConfigLib.readCore(vm, local);
+        
+        string memory key = uint256(coreConfig.digest()).toString();
+        string memory out = string.concat("deployments/core/", key);
+        try vm.readFileBinary(out) returns (bytes memory data) {
```
Comment (asaj) [inline scripts/DeployCore.s.sol:26]: Oh dope, does this work? Not sure why it didn't for me

File: scripts/DeployCore.s.sol [hunk 14]
```diff
@@ -3,31 +3,35 @@ pragma solidity ^0.8.17;
...
-        vm.startBroadcast();
-        uint256 startBlock = block.number;
+    function run() public returns (DeployLib.Core memory core) {
+        string memory local = vm.envString("LOCAL");
 
-        config.deploy(ismConfig);
-        config.check(ismConfig);
+        ConfigLib.Core memory coreConfig = ConfigLib.readCore(vm, local);
+        
+        string memory key = uint256(coreConfig.digest()).toString();
```
Comment (asaj) [inline scripts/DeployCore.s.sol:24]: Ohhh, is this what replaced incremental deploys?

File: scripts/DeployMultisigIsm.s.sol [hunk 15]
```diff
@@ -4,29 +4,42 @@ pragma solidity ^0.8.17;
...
+        address owner = vm.envAddress("OWNER");
+
+        ConfigLib.Multisig memory config = ConfigLib.readMultisigIsmConfig(
+            vm,
+            remotes,
+            owner
+        );
 
-        vm.startBroadcast();
+        string memory key = uint256(config.digest()).toString();
```
Comment (asaj) [inline scripts/DeployMultisigIsm.s.sol:32]: nit: Does this mean we can't deploy two MultisigIsms with the same config to two different chains?

File: scripts/DeployMultisigIsm.s.sol [hunk 16]
```diff
@@ -4,29 +4,42 @@ pragma solidity ^0.8.17;
...
+        string memory out = string.concat("deployments/ism/", key);
+        try vm.readFileBinary(out) returns (bytes memory data) {
+            ism = abi.decode(data, (MultisigIsm));
+        } catch {
+            vm.startBroadcast();
+            ism = config.deploy();
+            vm.stopBroadcast();
+            vm.writeFileBinary(out, abi.encode(ism));
```
Comment (asaj) [inline scripts/DeployMultisigIsm.s.sol:40]: Why binary, as opposed to a hex string?

File: scripts/DeployCore.s.sol [hunk 17]
```diff
@@ -3,31 +3,35 @@ pragma solidity ^0.8.17;
...
+        string memory out = string.concat("deployments/core/", key);
+        try vm.readFileBinary(out) returns (bytes memory data) {
+            core = abi.decode(data, (DeployLib.Core));
+        } catch {
+            vm.startBroadcast();
+            core = coreConfig.deploy();
+            vm.stopBroadcast();
+            vm.writeFileBinary(out, abi.encode(core));
```
Comment (asaj) [inline scripts/DeployCore.s.sol:32]: Why binary as opposed to json?

File: scripts/SendTestMessage.s.sol [hunk 18]
```diff
@@ -1,44 +0,0 @@
-// SPDX-License-Identifier: UNLICENSED
```
Comment (asaj) [inline scripts/SendTestMessage.s.sol:1]: Accidental? We should keep this

File: scripts/CheckMessage.s.sol [hunk 19]
```diff
@@ -1,42 +0,0 @@
-// SPDX-License-Identifier: UNLICENSED
```
Comment (asaj) [inline scripts/CheckMessage.s.sol:1]: Accidental? We should keep this

File: scripts/DeployCore.s.sol [hunk 20]
```diff
@@ -3,31 +3,35 @@ pragma solidity ^0.8.17;
...
-            .readHyperlaneDomainConfig(vm, local);
-        ConfigLib.MultisigIsmConfig memory ismConfig = ConfigLib
-            .readMultisigIsmConfig(vm, remotes);
+    using DeployLib for ConfigLib.Core;
+    using ConfigLib for ConfigLib.Core;
+    using CheckLib for DeployLib.Core;
 
-        vm.startBroadcast();
-        uint256 startBlock = block.number;
+    function run() public returns (DeployLib.Core memory core) {
```
Comment (asaj) [inline scripts/DeployCore.s.sol:19]: Note you'll need to modify the docs instructions now that multisig deployment does not happen as part of this (among other changes)

**Comment by asaj** [comment]
Also can we separate this from the changes to the MultisigIsm?

**Comment by asaj** [comment]
We've moved on from foundry

**Pull request closed without merging**